cnxauthoring/utils.py

```diff
@@ -221,23 +221,36 @@ def derive_resources(request, document):
 def profile_to_user_dict(profile):
     """Take a profile from openstax accounts and transform it into a local user
     format"""
-    # in case it's already in the local user format, no need to transform
-    if 'email' in profile:
-        return profile
-    email = None
-    for contact_info in profile.get('contact_infos') or []:
-        if contact_info.get('type') == 'EmailAddress':
-            email = contact_info.get('value')
-    firstname = profile.get('first_name') or ''
-    surname = profile.get('last_name') or ''
-    return {
-            'firstname': firstname,
-            'surname': surname,
-            'email': email or '',
-            'id': profile.get('username') or '',
-            'fullname': profile.get('fullname',
-                u'{} {}'.format(firstname, surname).strip()),
+    try:
+        user_profile = {
+            # A username or id MUST be supplied.
+            'id': profile.get('username') and profile['username'] or profile['id'],
             }
+    except KeyError:
+        raise ValueError("A 'username' or 'id' MUST be supplied "
+                         "in the profile argument.")
+    profile_attrs = [
+        # (<authoring-key>, <accounts-key>,),
+        ('firstname', 'first_name',),
+        ('surname', 'last_name',),
+        # ('fullname', 'fullname',),
+        ('suffix', 'suffix',),
+        ('title', 'title',),
+        ]
+    for au_key, acc_key in profile_attrs:
+        if au_key in profile:
+            user_profile[au_key] = profile[au_key]
+        else:
+            user_profile[au_key] = profile.get(acc_key, None)
+    if 'fullname' in profile:
+        user_profile['fullname'] = profile['fullname']
+    else:
+        firstname = user_profile['firstname']
+        surname = user_profile['surname']
+        fullname = u' '.join([n for n in (firstname, surname) if n])
+        user_profile['fullname'] = fullname or None
```

Comment:
The userrole schema ensures that, at the very least, the response will contain keys for 'firstname', 'surname' and 'fullname'. The 'title' and 'suffix' are dropped when they have no value.

The major difference here is the use of ``None`` (aka ``null``) instead of a blank string (``""``) for values that are empty. See also, openstax/accounts#128

Also, as you pointed out in Connexions/openstax-accounts#16, it is perfectly fine for accounts to respond with a profile that only has an 'id' and 'username'.